### PR TITLE
shud: handle websocket concurrency better

### PR DESCRIPTION
**Pull request by nicks**
Hello @maiamcc,

Please review the following commits I made in branch nicks/ch2404/concurrent:

cc6e31eab5afb58a705faba19ff4832e3f12e4c4 (2019-04-30 19:50:57 -0400)
shud: handle websocket concurrency better

### CODE REVIEW

File: internal/hud/server/websocket_test.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+		close(done)
+	}()
+
+	st.NotifySubscribers(ctx)
+
+	m := conn.AssertNextWriteMsg(t)
+
+	// Send a read error, and make sure the connection
+	// doesn't close immediately.
+	conn.readCh <- fmt.Errorf("read error")
```
Comment (maiamcc) [inline internal/hud/server/websocket_test.go]: oh, neat!!

File: internal/hud/server/websocket_test.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+type msg struct {
+	callback chan error
+}
+
+func (m msg) Ack() {
+	m.callback <- nil
+	close(m.callback)
+}
+
+func (m msg) Reject() {
```
Comment (maiamcc) [inline internal/hud/server/websocket_test.go]: you don't call this func at all -- is there supposed to be a test involving it?

File: internal/hud/server/websocket.go [hunk 3]
```diff
@@ -72,5 +84,13 @@ func (s HeadsUpServer) ViewWebsocket(w http.ResponseWriter, req *http.Request) {
 	ws := NewWebsocketSubscriber(conn)
 
 	// TODO(nick): Handle clean shutdown when the server shuts down
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/hud/server/websocket.go]: nope!

File: internal/hud/server/websocket.go [hunk 4]
```diff
@@ -72,5 +84,13 @@ func (s HeadsUpServer) ViewWebsocket(w http.ResponseWriter, req *http.Request) {
...
-	ws.Stream(context.TODO(), s.store)
+	ctx := context.TODO()
+
+	// Fire a fake OnChange event to initialize the connection.
+	ws.OnChange(ctx, s.store)
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/hud/server/websocket.go]: nope, because SetUp doesn't have access to store

File: internal/hud/server/websocket_test.go (hunk 2 above)
Comment (nicks) [reply internal/hud/server/websocket_test.go]: removed!

**Pull request merged**